bionic/code_references.py

```diff
@@ -0,0 +1,248 @@
+# Copyright 2018-2020 Streamlit Inc.
+# Author: Dominik Moritz
+
+# The code in this module is copied from Streamlit with modifications.
+# Unlike Streamlit, we don't import modules to avoid runtime
+# computation and use newer libraries to make the code cleaner.
+# https://github.com/streamlit/streamlit
+
+"""
+This module contains functions which are used to find references in
+your code objects. `get_referenced_objects` function finds those
+references for you given a code object and an additional `CodeContext`
+input, which can be extracted using the `get_code_context` function.
+
+CodeHasher uses these methods as part of hashing an entity function for
+Bionic cache invalidation. Using these methods, CodeHasher can hash the
+references of entity functions (and even their references), so that any
+change in the references are also detected by Bionic to invalidate
+cache.
+"""
+
+import attr
+import dis
+import inspect
+import warnings
+
+from .utils.misc import oneline
+
+
+@attr.s
+class CodeContext:
+    """
+    Holds variable information from function and code object
+    attributes.
+
+    Attributes
+    ----------
+    globals: dict
+        A reference to the dictionary that holds the function’s global
+        variables — the global namespace of the module in which the
+        function was defined.
+    cells: dict
+        A dictionary that tracks all free variables by name and cell
+        variable names by themselves.
+    varnames: dict
+        A dictionary that is used to track local variables by name.
+    can_find_references: bool, default True
+        Determines whether we can find references for the code object
+        using this reference.
+    """
+
+    globals = attr.ib(default={})
+    cells = attr.ib(default={})
+    varnames = attr.ib(default={})
+    can_find_references = attr.ib(default=True)
```

Comment:
I don't think there is anything useful we can do with the context when it is false. I was leaning towards keeping it None and asserting the value but this looked a bit cleaner when I tried. I can move back to None and assert though. Wdyt?

Reply:
It seems to me that there are two separate issues:

1. If we don't have a valid code context, should we represent that with `None` or with a special flag on the context object? I think `None` is preferable to having a flag that means "don't use this object for anything".
2. If we don't have a valid code context when processing a code object, should we throw an exception or just skip analyzing the code? I'm not sure what the right answer is. What do you think?

Reply:
1. I was leaning towards `None` too but this seemed a bit cleaner in case we want to ever change it later. I can go back to `None`.

2. I'm on the fence for this one too. I went from throwing an exception to ignoring it, but we won't ever get to this point unless someone passes a code object to the hasher directly (which shouldn't happen) or defines a code object themselves (do people even do this?). I can be easily convinced otherwise, it is totally reasonable to throw an exception as a guard against any future bugs. But if you don't feel strongly, I think I'll keep ignoring it in case anyone ever defines a code object by themself.